thirdparty/notifier/notifier.go

```diff
@@ -0,0 +1,113 @@
+//  type RocketNotifiee interface{
+//    Countdown(r Rocket, countdown time.Duration)
+//    LiftedOff(Rocket)
+//    ReachedOrbit(Rocket)
+//    Detached(Rocket, Capsule)
+//    Landed(Rocket)
+//  }
+//
+type Notifiee interface{}
+
+// Notifier is a notification dispatcher. It's meant
+// to be composed, and its zero-value is ready to be used.
+//
+//  type Rocket struct {
+//    notifier notifier.Notifier
+//  }
+//
+type Notifier struct {
+	mu   sync.RWMutex // guards notifiees
+	nots map[Notifiee]struct{}
+}
+
+// Notify signs up Notifiee e for notifications. This function
+// is meant to be called behind your own type-safe function(s):
+//
+//   // generic function for pattern-following
+//   func (r *Rocket) Notify(n Notifiee) {
+//     r.notifier.Notify(n)
+//   }
+//
+//   // or as part of other functions
+//   func (r *Rocket) Onboard(a Astronaut) {
+//     r.astronauts = append(r.austronauts, a)
+//     r.notifier.Notify(a)
+//   }
+//
+func (n *Notifier) Notify(e Notifiee) {
+	n.mu.Lock()
+	if n.nots == nil { // so that zero-value is ready to be used.
+		n.nots = make(map[Notifiee]struct{})
+	}
+	n.nots[e] = struct{}{}
+	n.mu.Unlock()
+}
+
+// StopNotifying stops notifying Notifiee e. This function
+// is meant to be called behind your own type-safe function(s):
+//
+//   // generic function for pattern-following
+//   func (r *Rocket) StopNotify(n Notifiee) {
+//     r.notifier.StopNotify(n)
+//   }
+//
+//   // or as part of other functions
+//   func (r *Rocket) Detach(c Capsule) {
+//     r.notifier.StopNotify(c)
+//     r.capsule = nil
+//   }
+//
+func (n *Notifier) StopNotify(e Notifiee) {
+	n.mu.Lock()
+	if n.nots != nil { // so that zero-value is ready to be used.
+		delete(n.nots, e)
+	}
+	n.mu.Unlock()
+}
+
+// NotifyAll messages the notifier's notifiees with a given notification.
+// This is done by calling the given function with each notifiee. It is
+// meant to be called with your own type-safe notification functions:
+//
+//  func (r *Rocket) Launch() {
+//    r.notifyAll(func(n Notifiee) {
+//      n.Launched(r)
+//    })
+//  }
+//
+//  // make it private so only you can use it. This function is necessary
+//  // to make sure you only up-cast in one place. You control who you added
+//  // to be a notifiee. If Go adds generics, maybe we can get rid of this
+//  // method but for now it is like wrapping a type-less container with
+//  // a type safe interface.
+//  func (r *Rocket) notifyAll(notify func(Notifiee)) {
+//    r.notifier.NotifyAll(func(n notifier.Notifiee) {
+//      notify(n.(Notifiee))
+//    })
+//  }
+//
+func (n *Notifier) NotifyAll(notify func(Notifiee)) {
+	n.mu.Lock()
+	if n.nots != nil { // so that zero-value is ready to be used.
+		for notifiee := range n.nots {
+			// we spin out a goroutine so that whatever the notification does
+			// it _never_ blocks out the client. This is so that consumers
+			// _cannot_ add hooks into your object that block you accidentally.
+			go notify(notifiee)
```

Comment:
> It's easy to detect leaked goroutines because the client will never receive an http response.

Not quite-- the http handler gets an io.Writer -- the client can write directly and then block forever. It can even flush it (io.Flusher). The HTTP server will continue merrily handling other requests. even more blocking ones. Example: http://gateway.ipfs.io/ipfs/QmR1XseVfGgcK4C7mqkikjJB8PFfAc1Mn2tm4vTh8F4Rxd

> I'm concerned about giving us rope to hang ourselves. This is our state of affairs:

Yeah, it's pretty bad. we need to fix that. We should be gathering stats + a distribution of _what_ goroutines it is. Any ideas on how to do that? maybe collecting metrics on all stacks every minute? 

And yeah, i'm open to suggestions. Do you have a simple rate limiting in mind?

Let's keep in mind this is a notification. the notificator _must not_ block on a notifiee being stupid. so if you want to add rate limiting of some sort, we'll need to do it without blocking the notifier.[1] (one possible approach is to put each notifiee in its own goroutine, so if it blocks it doesnt get more notifs. but this complicates the implementation. ). I dont think this is worth it, but if you think this is a serious problem, sure.

[1] I would not implement a notification structure that could have the side effect of giving the client hooks to break the underlying system. that would be like opening a hole in a server so clients can block it from making forward progress on anything else.